src/bp/core/services/converse.ts

```diff
@@ -0,0 +1,132 @@
+      order: 10000,
+      direction: 'outgoing',
+      handler: (event, next) => {
+        this._handleCapturePayload(event)
+        next()
+      }
+    })
+    this.eventEngine.register({
+      name: 'converse.capture.nlu',
+      description: 'Captures the nlu output for the Converse API',
+      order: 10000,
+      direction: 'incoming',
+      handler: (event, next) => {
+        this._handleCaptureNlu(event)
+        next()
+      }
+    })
+  }
+
+  public async sendMessage(botId: string, userId: string, payload): Promise<any> {
+    if (!payload.text || !_.isString(payload.text) || payload.text.length > 360) {
+      throw new InvalidParameterError('Text must be a valid string of less than 360 chars')
+    }
+
+    const sanitizedPayload = _.pick(payload, ['text', 'type', 'data', 'raw'])
+
+    // We remove the password from the persisted messages for security reasons
+    if (payload.type === 'login_prompt') {
+      sanitizedPayload.data = _.omit(sanitizedPayload.data, ['password'])
+    }
+
+    if (payload.type === 'form') {
+      sanitizedPayload.data.formId = payload.formId
+    }
+
+    const { result: user } = await this.userRepository.getOrCreate('api', userId)
+
+    const incomingEvent = Event({
+      type: 'text',
+      channel: 'api',
+      direction: 'incoming',
+      payload,
+      target: userId,
+      botId
+    })
+
+    await this.eventEngine.sendEvent(incomingEvent)
+
+    const timeoutPromise = this._createTimeoutPromise(userId)
+    const donePromise = this._createDonePromise(userId)
+
+    return Promise.race([timeoutPromise, donePromise])
+  }
+
+  private _createDonePromise(userId) {
+    return new Promise(resolve => {
+      converseApiEvents.on('done', event => {
+        if (event.target === userId) {
+          const json = this.jsonMap.get(event.target)
+          return resolve(json)
+        }
+      })
+    })
+  }
+
+  private _createTimeoutPromise(userId) {
+    return new Promise(resolve => {
+      converseApiEvents.on('action.running', event => {
+        if (event.target === userId) {
+          // ignore
+        } else {
+          const wait = setTimeout(() => {
+            clearTimeout(wait)
+            resolve({ error: 'Request timed out!' }) // TODO: Use reject on promise and use an error
+          }, 1000)
+        }
+      })
+    })
+  }
+
+  private _handleCapturePayload(event) {
+    if (this.jsonMap.has(event.target)) {
+      let json = this.jsonMap.get(event.target)
+      json = {
+        ...json,
+        response: [..._.get(json, 'response', []), event.payload],
+        state: _.get(event, 'state', {})
+      }
+      this.jsonMap.set(event.target, json)
+    } else {
+      this.jsonMap.set(event.target, {})
+    }
+  }
+
+  private _handleCaptureNlu(event) {
+    if (this.jsonMap.has(event.target)) {
+      let json = this.jsonMap.get(event.target)
+      const nlu = _.get(event, 'nlu', {})
```

Comment:
lodash not required when checking a single node:  const nlu = event.nlu || {}

Reply:
nlu is not defined in the event type, so yeah lodash is alright in this case. Will add nlu to the event type

Reply:
what yann meant is that there's no deep path, i.e. lodash does nothing / adds overhead. `event.nlu || {}` is more concise and much much faster